modular/gater/admin_handler.go

```diff
@@ -413,3 +419,223 @@ func (g *GateModular) replicateHandler(w http.ResponseWriter, r *http.Request) {
+	}()
+
+	// ignore the error, because the recovery request only between SPs, the request
+	// verification is by signature of the RecoveryTask
+	reqCtx, newErr := NewRequestContext(r, g)
+	if newErr != nil {
+		log.CtxDebugw(context.Background(), "recoveryPrimaryHandler new reqCtx error", "error", newErr)
+	}
+
+	recoveryMsg, err = hex.DecodeString(r.Header.Get(GnfdRecoveryMsgHeader))
+	if err != nil {
+		log.CtxErrorw(reqCtx.Context(), "failed to parse recovery header",
+			"header", r.Header.Get(GnfdRecoveryMsgHeader))
+		err = ErrDecodeMsg
+		return
+	}
+
+	recoveryTask := gfsptask.GfSpRecoveryPieceTask{}
+	err = json.Unmarshal(recoveryMsg, &recoveryTask)
+	if err != nil {
+		log.CtxErrorw(reqCtx.Context(), "failed to unmarshal receive header",
+			"receive", r.Header.Get(GnfdReceiveMsgHeader))
+		err = ErrDecodeMsg
+		return
+	}
+
+	// check signature consistent
+	taskSignature := recoveryTask.GetSignature()
+	signatureAddr, pk, err := RecoverAddr(crypto.Keccak256(recoveryTask.GetSignBytes()), taskSignature)
+	if err != nil {
+		log.CtxErrorw(reqCtx.Context(), "failed to  get recover task address", "error:", err)
+		err = ErrSignature
+		return
+	}
+
+	if !secp256k1.VerifySignature(pk.Bytes(), crypto.Keccak256(recoveryTask.GetSignBytes()), taskSignature[:len(taskSignature)-1]) {
+		log.CtxErrorw(reqCtx.Context(), "failed to verify recovery task signature")
+		err = ErrSignature
+		return
+	}
+
+	// handle request from primary SP to recovery segment {
+	objectInfo := recoveryTask.GetObjectInfo()
+	if objectInfo == nil {
+		log.CtxErrorw(reqCtx.Context(), "recovery task params error")
+		err = ErrInvalidHeader
+		return
+	}
+
+	if objectInfo.RedundancyType != storagetypes.REDUNDANCY_EC_TYPE {
+		log.CtxErrorw(reqCtx.Context(), "recovery redundancy type only support EC")
+		err = ErrRecoveryRedundancyType
+		return
+	}
+
+	chainObjectInfo, bucketInfo, params, err := getObjectChainMeta(reqCtx, g.baseApp, objectInfo.ObjectName, objectInfo.BucketName)
+	if err != nil {
+		err = ErrInvalidHeader
+		return
+	}
+
+	redundancyIdx := recoveryTask.EcIdx
+	maxRedundancyIndex := int32(params.GetRedundantParityChunkNum()+params.GetRedundantDataChunkNum()) - 1
+	if redundancyIdx < int32(RecoveryMinEcIndex) || redundancyIdx > maxRedundancyIndex {
+		err = executor.ErrRecoveryPieceIndex
+		return
+	}
+
+	var pieceData []byte
+	if redundancyIdx >= 0 {
+		// recovery secondary SP
+		pieceData, err = g.recoverySecondarySP(reqCtx.Context(), chainObjectInfo, bucketInfo, recoveryTask, params, signatureAddr)
+		if err != nil {
+			return
+		}
+	} else {
+		// recovery primary SP
+		pieceData, err = g.recoveryPrimarySP(reqCtx.Context(), chainObjectInfo, bucketInfo, recoveryTask, params, signatureAddr)
+		if err != nil {
+			return
+		}
+	}
+
+	w.Write(pieceData)
+	log.CtxDebugw(reqCtx.Context(), "succeed to get one ec piece data")
+}
+
+func (g *GateModular) recoveryPrimarySP(ctx context.Context, objectInfo *storagetypes.ObjectInfo,
```

Comment:
"recoverSegmentPiece" instead of "recoveryPrimarySP"

Reply:
fixed